hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -942,17 +944,159 @@ private boolean isKeyEmpty(OmKeyInfo keyInfo) {
+    List<OmKeyInfo> keyList = new ArrayList<>();
+    if (StringUtils.isBlank(startKey)) {
+      return listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName,
+              keyPrefix, maxKeys, keyList);
+    } else {
+      return listKeysRecursivelyFromStartKeyV1(volumeName, bucketName, startKey,
+              keyPrefix, maxKeys, keyList);
+    }
+  }
+
+  private List<OmKeyInfo> listKeysRecursivelyFromKeyPrefixV1(String volumeName,
+      String bucketName, String keyPrefix, int maxKeys,
+      List<OmKeyInfo> keyList) throws IOException {
+
+    String startKey = "";
+    int countEntries = keyList.size();
+    List<OzoneFileStatus> keys = listOzoneFileStatusesV1(false,
+            startKey, countEntries, maxKeys, null, volumeName,
+            bucketName, keyPrefix, true, true);
+
+    for (OzoneFileStatus status : keys) {
+      OmKeyInfo omKeyInfo = status.getKeyInfo();
+      keyList.add(omKeyInfo);
+      if (status.isDirectory()) {
+        // add trailing slash to represent directory
+        String dirName = omKeyInfo.getKeyName();
+        if (!dirName.endsWith("/")) {
+          omKeyInfo.setKeyName(dirName + OM_KEY_PREFIX);
+        }
+        if (!StringUtils.equals(dirName, keyPrefix)) {
+          listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName, dirName,
+                  maxKeys, keyList);
+        }
+      }
+    }
+    return keyList;
+  }
+
+  private List<OmKeyInfo> listKeysRecursivelyFromStartKeyV1(String volumeName,
+      String bucketName, String startKey, String keyPrefix,
+      int maxKeys, List<OmKeyInfo> keyList) throws IOException {
+
+    if (startKey.startsWith(keyPrefix)) {
+      // Back and forth searching method. Here, it needs to move the startKey
+      // & keyprefix pointers one step backward, then list all the keys
+      // forward to these to pointers.
+
+      // For example, startKey="dir1/dir2/dir3/file3" and keyPrefix="dir1/".
+      // Iterate then move startkey and parentKeyPrefixPath positions till the
+      // given keyPrefix and lists all its sup paths...
+
+      // Iteration-1:
+      // startKeyPos="dir1/dir2/dir3/file3", parentKeyPrefix="dir1/dir2/dir3"
+      // Iteration-2:
+      // startKeyPos="dir1/dir2/dir3", parentKeyPrefix="dir1/dir2"
+      // Iteration-3:
+      // startKeyPos="dir1/dir2", parentKeyPrefix="dir1"
+      String parentKeyPrefixPath = OzoneFSUtils.getParentDir(startKey, true);
+      String startKeyPos = startKey;
+
+      while (!StringUtils.equals(parentKeyPrefixPath, keyPrefix)) {
+        forwardSearchAndListKeysFromStartKeyToKeyPrefixV1(volumeName,
+                bucketName, startKeyPos, parentKeyPrefixPath, maxKeys,
+                keyList);
+
+        startKeyPos = parentKeyPrefixPath;
+        parentKeyPrefixPath = OzoneFSUtils.getParentDir(parentKeyPrefixPath,
+                true);
+      }
+    } else {
+      // TODO: Handle startKey not starting with keyPrefix. Should we suport
+      //  this case?
```

Comment:
If startKey does not start with key prefix, we can just compare the their value.

* If key prefix already behind the startKey, this search equals to the listKeys(key prefix , startKey='') as all entries with key prefix should behind the startKey.
* If key prefix not behind the startKey,  the empty list expected to be returned since startKey behind all the prefix key entries.

Does this handling make the sense here?